lib/spack/spack/compilers/__init__.py

```diff
@@ -221,7 +313,11 @@ def all_compilers_config(scope=None, init_config=True):
     """Return a set of specs for all the compiler versions currently
     available to build with.  These are instances of CompilerSpec.
     """
-    return get_compiler_config(scope, init_config)
+    from_compilers_yaml = get_compiler_config(scope, init_config)
+    from_packages_yaml = get_compiler_config_from_packages(scope)
+    result = from_compilers_yaml + from_packages_yaml
+    key = lambda c: _compiler_from_config_entry(c["compiler"])
+    return list(llnl.util.lang.dedupe(result, key=key))
```

Comment:
There are still issues here with functions that have an `init_config` parameter. Since now the configuration is spread over two files, we must get `packages.yaml` first and set `init_config` to True only if there are no compilers defined in there.

This is a reproducer for an issue that user would get otherwise (when running `spack spec` compilers.yaml gets initialized, and causes duplicates warning later):
![Screenshot from 2024-03-11 07-52-04](https://github.com/spack/spack/assets/4199709/499ac65f-346f-4aaa-ada7-55b95274e1de)

Reply:
Another minor issue is when we try to find compilers, and show where they are defined:
![Screenshot from 2024-03-11 07-54-39](https://github.com/spack/spack/assets/4199709/7c37d590-53bd-4f96-a72e-8d18a29bd974)

If the compilers are defined in `packages.yaml` the message shown to users will lack any reference to the file (and seems to leave the sentence half-way).

Reply:
These issues have been solved as of https://github.com/spack/spack/pull/42016/commits/74263b2016dd8cbeca0caedb8130e23dd7e114f7